### ISSUE TITLE
Webpack bundle.js size

### ISSUE THREAD

**Issue by Jucesr** [airbnb/react-dates#874]
I am using react-dates in my app. I am trying to reduce the size of my bundle.js generated by webpack. I used the "webpack-bundle-analyzer" and I noticed React-dates module is including components I am not using in my app. 

I only imported SingleDatePicker and RangeDatePicker but I am getting all the components in bundle. I tried to use the webpack.IgnorePlugin but I was not able to get this done. 

To import I used
`import { SingleDatePicker } from 'react-dates';`

**Comment by majapw** [comment]
Hi! Have you tries deep importing the component directly, e.g.
```
import SingleDatePicker from 'react-dates/lib/components/SingleDatePicker';
```

The index for `react-dates` includes all of the exports which might be
contributing.

**Comment by Jucesr** [comment]
Hi @majapw I tried but still it's kind of big. It's around 350kb. My bundle has other files such as DarPicker.js, CalendarMonth.js . I think they are used by SingleDataPicker that is why webpack is adding them.

Correct me if I am wrong please.

**Comment by ljharb** [comment]
Perhaps this is the only thing adding moment? Moment can be large in webpack because i believe it grabs up all locales by default.

**Comment by majapw** [comment]
SingleDatePicker does use a number of other components under the hood... it's a fairly complex component. moment is also quite a big package. 

@lencioni might have some more thoughts on optimization.

**Comment by lencioni** [comment]
Indeed, Moment can easily blow up your bundle size, especially if you aren't being really careful about what it brings in. I think they have a plan to make it easier to do the right thing, but I'm not sure if it has landed yet.

In https://github.com/airbnb/react-dates/pull/791 we added an /esm build that you will want to hook up via a webpack alias, if you are on a new enough version of webpack (2+ IIRC). https://webpack.js.org/configuration/resolve/#resolve-alias That should give you smaller modules for react-dates and allow tree shaking to actually work for the react-dates code.

Beyond that, I think it would be impactful to have a way to be able to code split parts of react-dates, so you only bring in the parts that are needed for the initial render for your parent bundle, and then async load the rest as needed. I opened an issue for this here: https://github.com/airbnb/react-dates/issues/839 However, I am expecting a baby at any moment so it is unlikely that I'll be able to do any work on that any time soon. If you are looking for a way to contribute and help solve this problem for everyone, that would be a great issue to dig into.

**Comment by jeznag** [comment]
@lencioni could you provide more details on how to use /esm with webpack alias?

**Comment by lencioni** [comment]
@jeznag I think you would need to add something like this to your webpack's resolve.alias configuration:

```js
'react-dates': 'react-dates/esm',
```

More info: https://webpack.js.org/configuration/resolve/#resolve-alias

**Comment by jorgecuesta** [comment]
Hi @lencioni I do what you suggest about alias and get same result about big size on bundle on `react-moment-proptypes`. I try using both lib and esm for `import DayPickerSingleDateController from 'react-dates/lib/components/DayPickerSingleDateController';` and get same size result.
Something else I can try to reduce the biggest size?

**Comment by jbfm** [comment]
@jorgecuesta I've worked around your exact issue. The problem is `react-moment-proptypes` depends on an old version of moment, so in your bundle you'll get both the peer dependecy from `react-dates`, which you've installed, and the old version from `react-moment-proptypes`, giving you an insane bundle size.

I solved this by simply aliasing moment in my webpack config to the later version. This could cause issues with `react-moment-proptype`, but I haven't noticed anything yet, been running it in production for a few weeks.

So something like this would probably help you
```
resolve: {
	alias: {
		moment: path.resolve(__dirname, 'node_modules/moment')
	},
},
```

**Comment by majapw** [comment]
Hmm, `react-moment-proptypes` has moment as both a peer dep and a direct dep (and it's acceptable range is open, it's like `>=1.6.0`) so whatever you provide should satisfy both `react-dates` and `react-moment-proptypes`'s requirements. I wonder if there's something weird in the deduping process. @ljharb

**Comment by ljharb** [comment]
I'd try running `npm dedupe` and see if that resolves your issues.

**Comment by jbfm** [comment]
Alright, @ljharb @majapw, I checked out an old commit and dug into what caused my issue a bit more. At the time I had this issue I had moment installed as `~2.18.0` due to some other dependency, this seems to cause an issue with Yarn since `react-dates` now had an unmet peer dependency, and in turn making it think `react-moment-proptypes` had an unmet depenecy as well, not really sure why it goes that way, using NPM instead of Yarn didn't give this issue.

```warning " > react-dates@16.0.1" has unmet peer dependency "moment@^2.18.1".
warning "react-dates > react-moment-proptypes@1.5.0" has unmet peer dependency "moment@>=1.6.0".```

And from there Yarn installed a duplicate moment within `react-moment-proptypes`

Not sure if this helps @jorgecuesta... :)

**Comment by jorgecuesta** [comment]
Thanks guys @majapw @ljharb and @jbfm to take time to see this. 
I have this dependencies: `"react-dates": "16.2.1"`
and finding moment get this result:
```
├─ moment@2.11.1
├─ react-moment-proptypes@1.5.0
│  └─ moment@2.20.1
```
I use Yarn all the time, maybe I need install same `moment` version that used by `moment-prop-types`

I will try using `npm` instead and check updating version. I will let you know if something of that reduce bundle size in relation to `react-dates`

**Comment by jeznag** [comment]
I set up the esm alias and don't see a difference in my bundle size.

**Comment by pedroabreu** [comment]
Running into the same issue while just importing the util `toMomentObject`. 

Importing it like `import { toMomentObject } from "react-dates"` I get the following in the bundle-analyzer

<img width="395" alt="screen shot 2018-05-23 at 22 50 40" src="https://user-images.githubusercontent.com/1194888/40453078-ec44a758-5edb-11e8-934c-bdf6ea79a4b7.png">

I'm now looking at another app, where I use the `DatePickerRangeController` and I also suspect it's bundling the entire `react-dates`. I would expect that not to be the case on a single component import, not at least for the entire package (a big chunk of it perhaps).

I will try the suggested deep importing

**Comment by mjmonline** [comment]
for moment optimizations try this plugin to strip redundant locales https://github.com/iamakulov/moment-locales-webpack-plugin

**Comment by Marvin1003** [comment]
@mjmonline you can pretty much achieve the same using webpacks Context Replacement Plugin

```
plugins: [
  new webpack.ContextReplacementPlugin(
    /moment[\/\\]locale$/,
    /de|fr|hu/
  )
]
```
https://webpack.js.org/plugins/context-replacement-plugin/

**Comment by jondavidjohn** [comment]
I built a tool to help you analyze webpack bundles for size regressions, and report them directly to GitHub PRs. It's free for open source, so it might be worth checking out and helpful in this scenario.

[https://packtracker.io](https://packtracker.io/?utm_campaign=links&utm_source=github&utm_medium=issueprcomment)

![image](https://user-images.githubusercontent.com/401520/58719965-8ae35c80-8396-11e9-908d-5c5b1d7b8cbf.png)

**Comment by thibaudcolas** [comment]
I think another optimisation that would greatly reduce the bundle size for Webpack users is to use the [`sideEffects`](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free) flag in the package.json to mark which files contain side effects.

At the moment, without it being set, even when using `react-dates/esm`, Webpack will bundle the whole library instead of just the selected imports. The workaround for this is to import components directly from the file where they are defined as https://github.com/airbnb/react-dates/issues/874#issuecomment-348569074 suggests, but is dependent on implementation details which are prone to breakage.

Additionally I think it would also make sense to declare the [`module`](https://webpack.js.org/guides/author-libraries/#final-steps) field in package.json pointing at `./esm/index.js`, since as far as I can see it’s indeed compiled with the same Babel config as the CJS build except for `modules: false`.
